### PR TITLE
[CBR-437] Kernel.Wallets.updatePassword should record if the user decided to remove it

### PR DESCRIPTION
**Pull request by Anviking**
## Description
When the password is updated we should check if it is the
`emptyPassphrase` and, if so, update `HdRoot` to have no spending
password.

This <s>commit</s> pr also adds a test property that ensures the hdRoot has no
spending password afterwards, and modifies the existing property to only
test updates with non-empty passwords, where we expect the new hdRoot to
*have* a spending password.

## Linked issue

[CBR-437](https://iohk.myjetbrains.com/youtrack/issue/CBR-437)

## Type of change

- [x] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🛠 New feature (non-breaking change which adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🏭 Refactoring that does not change existing functionality but does improve things like code readability, structure etc
- [ ] 🔨 New or improved tests for existing code
- [ ] ⛑ git-flow chore (backport, hotfix, etc)

## Developer checklist

- [x] I have read the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) document, and my code follows the code style of this project.
- [ ] If my code deals with exceptions, it follows the [guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md).
- [ ] I have updated any documentation accordingly, if needed. Documentation changes can be reflected in opening a PR on [cardanodocs.com](https://github.com/input-output-hk/cardanodocs.com), amending the inline [Haddock](https://www.haskell.org/haddock/) comments, any relevant README file or one of the document listed in the [docs](https://github.com/input-output-hk/cardano-sl/tree/develop/docs) directory.
- [ ] CHANGELOG entry has been added and is linked to the correct PR on GitHub.

## Testing checklist

- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: wallet-new/test/unit/Test/Spec/Wallets.hs [hunk 1]
```diff
@@ -311,8 +313,9 @@ spec = describe "Wallets" $ do
         describe "Wallet update password (kernel)" $ do
             prop "correctly replaces the ESK in the keystore" $ withMaxSuccess 50 $
                 monadicIO $ do
-                    newPwd <- pick arbitrary
+                    newPwd <- pick $ arbitrary `suchThat` (/= emptyPassphrase)
```
Comment (adinapoli-iohk) [inline wallet-new/test/unit/Test/Spec/Wallets.hs]: I would suggest not to touch the logic of this test, as this properly tests only one thing, which is the keystore replacement. What am I missing? What's the rationale behind the change?

File: wallet-new/src/Cardano/Wallet/Kernel/Wallets.hs [hunk 2]
```diff
@@ -384,7 +384,11 @@ updatePassword pw hdRootId oldPassword newPassword = do
                                -- was set.
                                -- Fix this properly as part of [CBR-404].
                                lastUpdateNow <- InDb <$> getCurrentTimestamp
-                               let hasSpendingPassword = HD.HasSpendingPassword lastUpdateNow
+
+                               let hasSpendingPassword = if (newPassword == emptyPassphrase)
+                                                         then HD.NoSpendingPassword
+                                                         else HD.HasSpendingPassword lastUpdateNow
```
Comment (adinapoli-iohk) [inline wallet-new/src/Cardano/Wallet/Kernel/Wallets.hs]: This is the way to go. However, also let's add a comment on _why_ we do this, as it might not be totally obvious to the reader that actually the frontend does allow the removal of the spending password after the wallet has been created (I, for one, got to know it just today)

File: wallet-new/test/unit/Test/Spec/Wallets.hs (hunk 1 above)
Comment (Anviking) [reply wallet-new/test/unit/Test/Spec/Wallets.hs]: I basically duplicated this test, checked for `HasSpendingPassword` here, and `NoSpendingPassword` in the new one.

I considered re-calculating the `HasSpendingPassword` from the arbitrary password [here](https://github.com/input-output-hk/cardano-sl/pull/3621/commits/df5a3ddd077ced9c3b9a5b38299d58385c36b652#diff-c76387a32496e0a7d02ba0d70564f7d0R336) making sure it was the same as the one in the hdRoot, but that felt to implicit, and dependent on `arbitrary` picking both empty and non-empty passwords.

I could make another copy? (Also, should I strip away the checks already tested here?)

Comment (adinapoli-iohk) [reply wallet-new/test/unit/Test/Spec/Wallets.hs]: My (relatively small) concern is that the first property is called "correctly replaces the ESK in the keystore" and, as such, it should work regardless of which spending password we pick to begin with. Right now it feels like this first test is a bit misleading, this is why I have proposed to leave it be and simply add another one (like you did) to check for this specific concern only.

**Comment by adinapoli-iohk** [comment]
Looks great, thanks @Anviking !
